docs/spec/connection-handshake/L1_2.md

```diff
@@ -0,0 +1,264 @@
+We enumerate the names of these datagrams in the following code snippet.
+
+```golang
+type ConnHandshakeInit struct {  }
+type ConnHandshakeTry struct {  }
+type ConnHandshakeAck struct {  }
+type ConnHandshakeConfirm struct {  }
+```
+
+3. A `StateProof` is a data type that abstracts over the details of a cryptographic proof that any process can create, and another process can verify.
+Intuitively, a proof is helpful for some process `x` to guarantee to another process `y` that `x` is in a certain state.
+Typically, a `StateProof` object is valid with respect to a `State` object.
+
+The diagram below captures the handshake protocol at a high level, comprising the four datagram exchanges.
+Note the direction of each arrow, indicating which is the sender and the destination process for each datagram, respectively.
+
+```
+               The four messages comprising
+             IBC Connection Handshake Protocol
+
+     +-----------+   ConnHandshakeInit   +-----------+
+     |           |--------- 1 --------->>|           |
+     |           |                       |           |
+     |           |   ConnHandshakeTry    |           |
+     |           |<<------- 2 -----------|           |
+     | Initiator |                       | Responder |
+     |     I     |   ConnHandshakeAck    |     R     |
+     |           |--------- 3 --------->>|           |
+     |           |                       |           |
+     |           | ConnHandshakeConfirm  |           |
+     |           |<<------- 4 -----------|           |
+     +-----------+                       +-----------+
+```
+
+
+###### Protocol handlers
+
+```golang
+// This function handles the request to execute a connection handshake between
+// process `i` (the initiator) and remote process (the responder) `r`.
+func ConnHandshakeInit(
+  identifier Identifier,
+  clientIdentifier Identifier, counterParty CounterParty) {
+    if connectionEnd != NilConnection then error
+
+    connectionEnd.connectionIdentifier = identifier
+    connectionEnd.clientIdentifier = clientIdentifier
+    connectionEnd.state = INIT
+    connectionEnd.counterParty = counterParty
+
+    consensusStateB = getConsensusState(connectionEnd.clientIdentifier)
+
+    datagram = ConnOpenTry(
+    connectionEnd, getProof(connectionEnd),
+    consensusStateB, getProof(consensusStateB))
+    send (datagram, r)
+}
+```
+
+```golang
+func ConnHandshakeTry(
+  connectionEnd_A Type, proofOfConnectionEnd Type,
+  consensusStateB Type, proofOfConsensusStateB Type) {
+    if connectionEnd != NilConnection then error
+
+verifyConnectionState(proofOfConnectionEnd, connectionEnd_A) verifyClientConsensusState(proofOfConsensusStateB, consensusStateB)
+
+connectionEnd = createConnectionEnd(connectionEnd_A)
+connectionEnd.state = TRYOPEN
+
+consensusStateA = getConsensusState(connectionEnd.clientIdentifier)
+
+send ConnOpenAck( connectionEnd, getProof(connectionEnd), consensusStateA, getProof(consensusStateA))
+}
+```
+
```

Comment:
The code is not aligned, missing parenthesis, etc. It is not clear on what side (A vs B) these are run. I think code  and figure above are inconsistent.